### DISCUSSION TITLE
What would be best practice to modify extensions?

### DISCUSSION THREAD

**Discussion by JetGum** [datenstrom/community#532] [Ask a question] [answered]
Hi all,

I have modified a couple of extensions to my needs while the original extension name has remained the same. These changes obviously get overwritten whenever extension is updated.

Question: What would be best practice to keep my changes **and** use updated (and improved) original extension code? 

One idea that came to mind: 

a) Copy modified extension code into a new .php file and rename it so it is still identifiable (eg: breadcrumbs.php > brdcrmbs.php). 
b) Change shortcodes on website ([breadcrumbs] > [brdcrmbs])
c) Evaluate and compare my modified extension on code line level whenever an update is available
d) Do a-c for multiple projects

Does that make sense to you? Does that even work?

Cheers,

Carsten

**Comment by markseuffert** [comment]
Hello Carsten, here are some options:

1. Send a pull request to the developer, it may or may not be accepted.
2. Discuss your modifications with the Datenstrom community.
3. Make a new extension with the modified code.

Here are some tips for handling modified code:

- Pick a unique file name. For example `breadcrumbc.php`, here the letter `c` stands for Carsten. When experimenting with new features adding a letter/postfix to the file name is a quick solution. I do this sometimes, so that original and modified code will be shown side-by-side in the file system.
- Use the same name in class/shortcut/settings. For example use `class YellowBreadcrumbc` and shortcut `[breadcrumbc]`. This is one of the software patterns that you can find in many extensions. That's how we avoid name collisions and it also makes it super easy to understand how different pieces fit together. Just by looking at the name.
- Keep the original attribution in each source file. For example: `// Based on Breadcrumb extension, http://website`. You are probably used to give recognition to other people's work if you are working with open source software. A good practice is to mention other people, libraries and services in the documentation as well. 
- Publish your changes. For example upload it to GitHub or Codeberg. While it's recommended to put extensions into their own repository, you could collect small changes in one repository. The idea is that I give you source code and you give me your changes back, [as explained in this video](https://www.youtube.com/watch?v=PaKIZ7gJlRU). This is one of the reasons why we chose the [GPL version 2](https://github.com/datenstrom/yellow/blob/main/LICENSE.md).

Hope this helps and happy coding.

**Comment by schulle4u** [comment, marked as answer]
Hello Carsten, 

an even better practice would be to suggest your modifications as possible new features to the developers, perhaps you will then have less trouble updating your sites. 😀
But I think your strategy should be OK. I have done the same with the Feed extension and extended/renamed it for [podcast publishing](https://github.com/schulle4u/yellow-extensions-schulle4u/tree/master/podcast). I try to keep track of the original extension code and apply API and layout changes to my modified extension. 

Steffen

**Comment by JetGum** [comment]
Hi guys,

thx for your answers. This clarifies things for me!

From my perspectivem my modifications would not require the creation of a new extension, for I am usually not adding much code or functionality but rather comment and replace a couple of minor things (mainly with regard to SEO or Page speed). Standing on the shoulders of giants, so to speak. Example: In toc.php, I have commented out a couple of lines to blend out 2nd/3rd level headings. And I have changed the code slightly so the numbering of the headings would not be included in the toc links. Thereforem I would refrain from creating a separate extension, but would opt for what I have initially proposed (rename locally, and check from time 2 time the original code). 

On top of that and in cases when there is substantial modification created, I would send a pull request and/or contact the developer. Here, I agree with you,

Cheers,

Carsten
